src/JSONParseErrorException.php

```diff
@@ -0,0 +1,49 @@
+<?php
+/*
+    Copyright 2016 Rustici Software
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+namespace TinCan;
+
+class JSONParseErrorException extends \Exception
+{
+    private static $format = 'Invalid JSON "%s": %s (%d)';
+
+    private $malformedValue;
+    private $jsonErrorNumber;
+    private $jsonErrorMessage;
+
+    public function __construct($malformedValue, $jsonErrorNumber, $jsonErrorMessage, \Exception $previous = null) {
+        $this->malformedValue   = $malformedValue;
+        $this->jsonErrorNumber  = (int) $jsonErrorNumber;
+        $this->jsonErrorMessage = $jsonErrorMessage;
```

Comment:
Right, it isn't leaky because we could be acting on the internal variables in some way and the interface is maintained. Perhaps "encapsulation" isn't the best term for what is being described, but calling the json_* functions for error handling inside is most certainly a side effect. The reason we have to couple the json_* calls is to prevent the possibility that we lose data by having multiple calls to json_* functions not happening sequentially. By example:

> json_decode(...);
> json_decode(.......);
> new JsonParse...Exception();

Now in the construction of the exception you've lost the details from the first call to json_decode. Frankly this is just a problem in PHP's interface, which ought to be atomic, that we shouldn't have to correct for, but here we are.